### PR TITLE
#21475 moving compressions ownership to Decoder

### PR DESCRIPTION
**Pull request by ktoso**
Work in progress on #21475

Sometimes the integration spec now is 2 off and I'm thinking if that could be correct or not... A newer advertisement than expected technically could be valid still. It's not originating in the removal of atomics, had this failure once before.

If someone wants to give initial feedback please do, though it's not all cleaned up yet.

For example I'm thinking if we need to do the `done.success(Done)` dance, since the intialization IMO always will be completed before we're hit with this so the timeout guards are not really needed. And missing an ACK is OK as well, since it would be re-delivered - shall I remove the Done dances from the `...Cb` callbacks then? I don't use the futures anyway now.

### CODE REVIEW

File: akka-remote/src/main/scala/akka/remote/artery/Codecs.scala
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/Codecs.scala]: hm, this requires some thinking. What happens if the inbound stream is restarted?
Then messages compressed with the old table may arrive, in best case we can't decompress them, in worse case it's a matching table version with different content and decompressed to wrong thing.

I'm not sure if this is a real problem.

If it is we need to create the instance in `ArteryTransport`.

**Review by patriknw — COMMENTED** [review]
looking great, first round of comments...

File: akka-remote/src/main/scala/akka/remote/artery/ArterySettings.scala
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/artery/ArterySettings.scala]: true thanks

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: good

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: I've split it

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: removing, some of those still laying around from when I wasn't sure if we need this or not.
With the move of compression creation to the outside I think we do need it again, so uncommenting actually :)

File: akka-remote/src/main/scala/akka/remote/artery/compress/InboundCompressions.scala
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/artery/compress/InboundCompressions.scala]: Nope, it's not thread safe.

Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/artery/compress/InboundCompressions.scala]: Removing all this dance in any case then

File: akka-remote/src/main/scala/akka/remote/artery/compress/InboundCompressions.scala
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/artery/compress/InboundCompressions.scala]: has to be reated here from the system anyway, once per creation of `InboundCompressionsImpl` not sure we can work around that?

Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/artery/compress/InboundCompressions.scala]: Oh my bad, yes can be shared.

File: akka-remote/src/main/scala/akka/remote/artery/Codecs.scala
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/artery/Codecs.scala]: Yeah you're right, that could cause trouble. I've worked on a test and now attempting to get it to fail and refactor the creation of the table

**Comment by ktoso** [comment]
Pushed rework as: https://github.com/akka/akka/pull/22047/commits/78ecb83fd8d2c667fd39ee67871688b0921f85c4

File: akka-remote/src/main/scala/akka/remote/artery/Association.scala
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/artery/Association.scala]: I made it an OptionVal some time ago, so foreach is not an option. You're right it's not hot-patch. Should I revert it to Option?

Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/Association.scala]: ah, I missed that. Not important, keep it as is.

File: akka-remote/src/main/scala/akka/remote/artery/Codecs.scala
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/artery/Codecs.scala]: second opinion on that one? The callbacks in the Encoder are wrapped like that since it is initialized later, however in the Decoder once constructed we're ready to go so I don't see how the invokations could fail.
On the other hand, it's not hot path so it may be safer this way in case something goes wrong.
We don't use the futures anywhere though. So I was tempted to not use any of the Future dance (see like 274 for an example)

File: akka-remote/src/main/scala/akka/remote/artery/Codecs.scala
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/artery/Codecs.scala]: AFAICS this should never really happen for the Decoder

File: akka-remote/src/main/scala/akka/remote/artery/Codecs.scala
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/artery/Codecs.scala]: So that's the tradeoff we're making. This used to be outside of hotpath, but now that we own the compression in this stage and don't synchronise it has to be done here.

File: akka-remote/src/main/scala/akka/remote/artery/FlightRecorderEvents.scala
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/artery/FlightRecorderEvents.scala]: Started adding some more AFR events.

File: akka-remote/src/main/scala/akka/remote/artery/compress/InboundCompressions.scala
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/artery/compress/InboundCompressions.scala]: @patriknw so the tombstones were added to avoid signalling a closed connection AFAIR, but now we have some garbage lingering around forever. I think next step here could be simply to schedule a cleanup callback once we put the None in?

File: akka-remote/src/main/scala/akka/remote/artery/compress/InboundCompressions.scala
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/artery/compress/InboundCompressions.scala]: does not have to be atomic anymore

File: akka-remote/src/main/scala/akka/remote/artery/compress/InboundCompressions.scala
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/artery/compress/InboundCompressions.scala]: notable change since previous logs were confusing - you'd think you have 2 compression instances which sounds scary, but actually is true and ok – one for manifests and one for actor refs, how the log message is better at explaining that.

File: akka-remote/src/main/scala/akka/remote/artery/compress/TopHeavyHitters.scala
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/artery/compress/TopHeavyHitters.scala]: Oh, right we can now remove this. On it.

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala
Comment (jrudolph) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: Conflict resolution looks good AFAICS.

File: akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/SurviveInboundStreamRestartWithCompressionInFlightSpec.scala
Comment (patriknw) [inline akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/SurviveInboundStreamRestartWithCompressionInFlightSpec.scala]: why is this needed?

File: akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/SurviveInboundStreamRestartWithCompressionInFlightSpec.scala
Comment (patriknw) [inline akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/SurviveInboundStreamRestartWithCompressionInFlightSpec.scala]: awaitAssert is better, no need for the Try and better error reporting

File: akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/SurviveInboundStreamRestartWithCompressionInFlightSpec.scala
Comment (patriknw) [inline akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/SurviveInboundStreamRestartWithCompressionInFlightSpec.scala]: perhaps use a shorter timeout for this `expectMsg`

File: akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/SurviveInboundStreamRestartWithCompressionInFlightSpec.scala
Comment (patriknw) [inline akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/SurviveInboundStreamRestartWithCompressionInFlightSpec.scala]: good test

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: I think it would be enough with `volatile var`, but is it even possible to create it in constructor and just use a `val`?

File: akka-remote/src/main/scala/akka/remote/artery/TestStage.scala
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/TestStage.scala]: to be strict this should be an atomic check/get/update, i.e. remove `shouldFailInbound` and return Option from `getInboundFailureOnce`

File: akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/SurviveInboundStreamRestartWithCompressionInFlightSpec.scala
Comment (ktoso) [inline akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/SurviveInboundStreamRestartWithCompressionInFlightSpec.scala]: not needed  true

File: akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/SurviveInboundStreamRestartWithCompressionInFlightSpec.scala
Comment (ktoso) [inline akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/SurviveInboundStreamRestartWithCompressionInFlightSpec.scala]: great, thanks :)

File: akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/SurviveInboundStreamRestartWithCompressionInFlightSpec.scala
Comment (ktoso) [inline akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/SurviveInboundStreamRestartWithCompressionInFlightSpec.scala]: done

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: True, it can be - done!

File: akka-remote/src/main/scala/akka/remote/artery/TestStage.scala
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/artery/TestStage.scala]: True good catch

File: akka-remote/src/main/scala/akka/remote/artery/compress/InboundCompressions.scala
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/artery/compress/InboundCompressions.scala]: have a look on this comment please @patriknw 
Would be done in separate PR

**Comment by ktoso** [comment]
Pushed final cleanups

**Review by patriknw — COMMENTED** [review]
good

File: akka-remote/src/main/scala/akka/remote/artery/Codecs.scala
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/Codecs.scala]: in theory it can happen, I think, so I think it's good to keep the catch

**Review by patriknw — APPROVED** [review]
LGTM, before merging I would like to run it a few times in a jenkins job that runs all multi-jvm tests. I'll do that when I'm done with my current testing.

**Comment by patriknw** [comment]
Started repeat job, but there seems to be some initialization problem:
```
[info] akka.remote.artery.RemoteDeployerSpec *** ABORTED *** (6 milliseconds)
[info]   scala.MatchError: null
[info]   at akka.remote.artery.ArteryTransport.createFlightRecorderEventSink(ArteryTransport.scala:385)
[info]   at akka.remote.artery.ArteryTransport.<init>(ArteryTransport.scala:321)
[info]   at akka.remote.RemoteActorRefProvider.init(RemoteActorRefProvider.scala:204)
[info]   at akka.actor.ActorSystemImpl.liftedTree2$1(ActorSystem.scala:807)
[info]   at akka.actor.ActorSystemImpl._start$lzycompute(ActorSystem.scala:804)
[info]   at akka.actor.ActorSystemImpl._start(ActorSystem.scala:804)
[info]   at akka.actor.ActorSystemImpl.start(ActorSystem.scala:820)
```

https://jenkins.akka.io:8498/job/akka-multi-node-repeat/12657/consoleText

**Comment by ktoso** [comment]
Hm weird I'll have a look

**Comment by ktoso** [comment]
Fixed, really weird... initialization order dance was required (moved some fields around).

**Comment by patriknw** [comment]
thanks, starting the repeat job on this branch again

**Comment by patriknw** [comment]
Some 20+ runs of akka-remote and akka-remote-tests yesterday. Two failures, but I think they are unrelated (not even related to Artery) and already known failures:

https://jenkins.akka.io:8498/job/akka-multi-node-repeat/12680
https://github.com/akka/akka/issues/21662

https://jenkins.akka.io:8498/job/akka-multi-node-repeat/12658/
https://github.com/akka/akka/issues/20180

I'll continue running it today but with all cluster tests with artery enabled.

**Comment by ktoso** [comment]
Cool, great :)

**Comment by patriknw** [comment]
The repeated cluster tests seems good, so can be merged. Maybe review by @drewhk first?

**Comment by patriknw** [comment]
@ktoso Let's merge this. Please squash

**Comment by ktoso** [comment]
Merged 👍

**Comment by ktoso** [comment]
I'll backport

**Pull request merged**
